src/main.cairo

```diff
@@ -0,0 +1,158 @@
+    use starknet::class_hash::ClassHash;
+    use ecdsa::check_ecdsa_signature;
+
+    use sol_subdomain_distribution::interface::sol_subdomain::ISolSubdomain;
+    use naming::interface::resolver::IResolver;
+    use openzeppelin::upgrades::{UpgradeableComponent, interface::IUpgradeable};
+
+    component!(path: UpgradeableComponent, storage: upgradeable, event: UpgradeableEvent);
+    impl UpgradeableInternalImpl = UpgradeableComponent::InternalImpl<ContractState>;
+
+    #[storage]
+    struct Storage {
+        identity_contract: ContractAddress,
+        naming_contract: ContractAddress,
+        admin: ContractAddress,
+        server_pub_key: felt252,
+        name_owners: LegacyMap::<felt252, ContractAddress>,
+        resolving_mapping: LegacyMap::<felt252, ContractAddress>,
+        #[substorage(v0)]
+        upgradeable: UpgradeableComponent::Storage,
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        DomainClaimed: DomainClaimed,
+        DomainResolvingUpdate: DomainResolvingUpdate,
+        #[flat]
+        UpgradeableEvent: UpgradeableComponent::Event
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct DomainClaimed {
+        #[key]
+        domain: Span<felt252>,
+        address: ContractAddress,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct DomainResolvingUpdate {
+        #[key]
+        domain: Span<felt252>,
+        field: felt252,
+        target_addr: ContractAddress,
+    }
+
+    #[constructor]
+    fn constructor(
+        ref self: ContractState,
+        admin_address: ContractAddress,
+        starknetid_address: ContractAddress,
+        naming_address: ContractAddress,
+        pub_key: felt252
+    ) {
+        self.identity_contract.write(starknetid_address);
+        self.naming_contract.write(naming_address);
+        self.admin.write(admin_address);
+        self.server_pub_key.write(pub_key);
+    }
+
+    #[external(v0)]
+    impl SolSubdomainImpl of ISolSubdomain<ContractState> {
+        fn claim(
+            ref self: ContractState, name: felt252, sig: (felt252, felt252), max_validity: u64,
+        ) {
+            assert(get_block_timestamp() < max_validity.try_into().unwrap(), 'Signature expired');
+
+            let caller = get_caller_address();
+            let message_hash: felt252 = hash::LegacyHash::hash(
+                hash::LegacyHash::hash(hash::LegacyHash::hash('sol subdomain', max_validity), name),
+                get_caller_address()
+            );
+
+            let public_key = self.server_pub_key.read();
+            let (sig0, sig1) = sig;
+            let is_valid = check_ecdsa_signature(message_hash, public_key, sig0, sig1);
+            assert(is_valid, 'Invalid signature');
+
+            self.name_owners.write(name, caller);
+
+            self
+                .emit(
+                    Event::DomainClaimed(
+                        DomainClaimed { domain: array![name, 16434].span(), address: caller, }
```

Comment:
Please don't emit the parent domain in the event. It's not a big deal in that example but this is good to standardize now. The idea is that in the future, if you create a custom resolver you can connect multiple domains to it. For example if you migrated iris.sol, you'll have access to iris.sol.stark and also an aliase iris.solana.stark without to write anything else on chain, because both domains would use the same resolver.